plugins/outputs/warp10/warp10.go

```diff
@@ -0,0 +1,204 @@
+package warp10
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+const (
+	defaultClientTimeout = 15 * time.Second
+)
+
+// Warp10 output plugin
+type Warp10 struct {
+	Prefix  string
+	WarpURL string
+	Token   string
+	Timeout internal.Duration `toml:"timeout"`
+	client  *http.Client
+}
+
+var sampleConfig = `
+  # prefix for metrics class Name
+  prefix = "Prefix"
+  ## POST HTTP(or HTTPS) ##
+  # Url name of the Warp 10 server
+  warp_url = "WarpUrl"
+  # Token to access your app on warp 10
+  token = "Token"
+  # Warp 10 query timeout, by default 15s
+  timeout = "15s"
+`
+
+// MetricLine Warp 10 metrics
+type MetricLine struct {
+	Metric    string
+	Timestamp int64
+	Value     string
+	Tags      string
+}
+
+func (w *Warp10) createClient() (*http.Client, error) {
+	client := &http.Client{
+		Timeout: w.Timeout.Duration,
+	}
+
+	return client, nil
+}
+
+// Connect to warp10
+func (w *Warp10) Connect() error {
+	client, err := w.createClient()
+	if err != nil {
+		return err
+	}
+
+	if w.Timeout.Duration == 0 {
+		w.Timeout.Duration = defaultClientTimeout
+	}
+
+	w.client = client
+	return nil
+}
+
+// GenWarp10Payload compute Warp 10 metrics payload
+func (w *Warp10) GenWarp10Payload(metrics []telegraf.Metric, now time.Time) string {
+	collectString := make([]string, 0)
+	index := 0
+	for _, mm := range metrics {
+
+		for _, field := range mm.FieldList() {
+
+			metric := &MetricLine{
+				Metric:    fmt.Sprintf("%s%s", w.Prefix, mm.Name()+"."+field.Key),
+				Timestamp: now.UnixNano() / 1000,
+			}
+
+			metricValue, err := buildValue(field.Value)
+			if err != nil {
+				log.Printf("Warp: %s\n", err.Error())
+				continue
+			}
+			metric.Value = metricValue
+
+			tagsSlice := buildTags(mm.Tags())
+			metric.Tags = strings.Join(tagsSlice, ",")
+
+			messageLine := fmt.Sprintf("%d// %s{%s} %s\n", metric.Timestamp, metric.Metric, metric.Tags, metric.Value)
+
+			collectString = append(collectString, messageLine)
+			index++
+		}
+	}
+	return fmt.Sprint(strings.Join(collectString, ""))
+}
+
+// Write metrics to Warp10
+func (w *Warp10) Write(metrics []telegraf.Metric) error {
+
+	if len(metrics) == 0 {
+		return nil
+	}
+	var now = time.Now()
+	payload := w.GenWarp10Payload(metrics, now)
+
+	req, err := http.NewRequest("POST", w.WarpURL, bytes.NewBufferString(payload))
+	req.Header.Set("X-Warp10-Token", w.Token)
+	req.Header.Set("Content-Type", "text/plain")
+
+	resp, err := w.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 {
+		body, _ := ioutil.ReadAll(resp.Body)
```

Comment:
I'd like to improve this error a bit, do you have an example of the response body on an error?

Reply:
**Case of an invalid token**:
> 2019-08-24T09:53:16Z E! [agent] Error writing to output [warp10]: http://127.0.0.1:8090: <html>
> <head>
> <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
> <title>Error 500 io.warp10.script.WarpScriptException: Invalid token.</title>
> </head>
> <body><h2>HTTP ERROR 500</h2>
> <p>Problem accessing /api/v0/update. Reason:
> <pre>    io.warp10.script.WarpScriptException: Invalid token.</pre></p>
> </body>
> </html>

**Case of an unvalid input**:
> 2019-08-24T09:53:16Z E! [agent] Error writing to output [warp10]: http://127.0.0.1:8090: <html>
> <head>
> <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
> <title>Error 500 Parse error at &apos;1566640396329990// telegraf{} 5 test.processes.idle{host=laptop1134849.home,source=telegraf} 0&apos;</title>
> </head>
> <body><h2>HTTP ERROR 500</h2>
> <p>Problem accessing /api/v0/update. Reason:
> <pre>    Parse error at &apos;1566640396329990// telegraf{} 5 test.processes.idle{host=laptop1134849.home,source=telegraf} 0&apos;</pre></p>
> </body>
> </html>

Reply:
I just added all error message I know about that can be returned by Warp10. As well as some test body examples. 
Warp is sending only 200 as return code in case of success, I replaced the test to use an equality check here.

Reply:
I'm not entirely satisfied with the error handling yet.  Using the docker container I poked around a bit at the API, and I noticed that the error string is always included in the response status line, here are few examples:
```
HTTP/1.1 500 Error when updating data: Invalid token.
HTTP/1.1 500 Error when updating data: Parse error at '/ test{} 42' (Missing location/elevation separator.) (Missing location/elevation separator.)
```

I also noticed that the status line does not use HTML escaping, which is a bit nicer especially in the case of an unexpected error.  I think we should use the status line as the source for the error strings.  We can split it into three pieces using strings.SplitN, the third part will be the error string.